source/agora/consensus/Ledger.d

```diff
@@ -827,15 +828,46 @@ public class Ledger
                 block_time_offset_tolerance))
             return reason;
 
+        // At this point we know it is the next block and also that it isn't Genesis
         try
         {
-            if (block.header.random_seed != this.getRandomSeed(block.header.height, block.header.missing_validators))
-                return "Block: Header's random seed does not match that of known pre-images";
+            const validators = this.getValidators(block.header.height);
+            if (validators.length != block.header.preimages.length)
+                return "Block: Number of preimages does not match active validators";
+            foreach (idx, const ref hash; block.header.preimages)
+            {
+                if (hash is Hash.init)
+                {
+                    // TODO: Check that the block contains a slashing transaction
+                }
+                // We don't have this pre-image yet
+                else if (validators[idx].preimage.height < block.header.height)
+                {
+                    PreImageInfo pi = validators[idx].preimage;
+                    pi.height = block.header.height;
+                    pi.hash = hash;
+                    if (!this.addPreimage(pi))
+                        return "Block: Preimages include an invalid non-revealed pre-image";
+                }
+                else
+                {
+                    // TODO: By caching the 'current' hash, we can prevent a semi
+                    // DoS if a node reveal a pre-image far in the future and then
+                    // keep on submitting wrong blocks.
+                    if (hash.hashFull() !is validators[idx].preimage[Height(block.header.height - 1)])
```

Comment:
Why is it hashing here ? :|

Reply:
Verifying the pre-image is valid.

Reply:
We could just compare them.

Reply:
Yes, as the check that it `hashed` is the previous is performed when receiving the `pre-image`.